Catch events in Azure DevOps Test Runner and forward them to Teamscale

rottkit opened the pull request:
… Teamscale

rottkit committed:
CR#19156 Catch events in Azure DevOps Test Runner and forward them to Teamscale

rottkit committed:
CR#19156 v0.3 as uploaded to Chrome Web Store

bahadirk commented on options.html:
```diff
@@ -0,0 +1,43 @@
+    <p>SAP Username (executes the test)</p>
+    <p><input id="sap-username" type="text" name="sap-username" class="ts-option"></p>
+
+    <p>Additional url filter to activate plugin on other Azure DevOps Server<br>
```
```suggestion
    <p>Additional URL filter to activate the plugin on the other Azure DevOps Server<br>
```

rottkit replied:
adjusted documentation

bahadirk commented on options.html:
```diff
@@ -0,0 +1,43 @@
+<h2>Configure Teamscale Server and SAP Username</h2>
+<div>
+    <p>Teamscale Server (note that you have to be logged in while testing)<br>
+        Please include protocol <span class="example-url">(e.g. http://teamscale.example.com</span>)</p>
```
```suggestion
        Please include a protocol <span class="example-url">(e.g. http://teamscale.example.com</span>)</p>
```

rottkit replied:
added `the`

bahadirk commented on options.js:
```diff
@@ -0,0 +1,27 @@
+const tsProjectOptionName = 'ts-project';
+const sapUserOptionName = 'sap-username';
+const extendedUriFilterOptionName = 'extended-uri-filter';
+const allOptions = [tsServerOptionName, tsProjectOptionName, sapUserOptionName, extendedUriFilterOptionName];
```
AFAIS, these `const` parameters are actually an `id`, right? Because we get these elements by using these `names` below by calling `getElementById()`. If so, maybe renaming them like would make much sense:

```suggestion
const tsServerOptionId = 'ts-server';
const tsProjectOptionId = 'ts-project';
const sapUserOptionId = 'sap-username';
const extendedUriFilterOptionId = 'extended-uri-filter';
const allOptionIds = [tsServerOptionId, tsProjectOptionId, sapUserOptionId, extendedUriFilterOptionId];
```

rottkit replied:
renamed with `id` suffix and moved to `constants.js`

bahadirk commented on options.js:
```diff
@@ -0,0 +1,27 @@
+
+function loadDataAndListen() {
+    chrome.storage.local.get(allOptions, result => {
+        allOptions.forEach(optionName => {
```
If the renaming above makes sense, then:
```suggestion
        allOptionIds.forEach(optionId => {
```

rottkit replied:
done

bahadirk commented on options.js:
```diff
@@ -0,0 +1,27 @@
+    chrome.storage.local.get(allOptions, result => {
+        allOptions.forEach(optionName => {
+            const optionInput = document.getElementById(optionName);
+            optionInput.value = result[optionName];
```
AFAIR, `getElementById()` can also return `null` or `undefined`. I am not sure whether it is possible in this context, but I think we should always add a `null` check to prevent getting `undefined` error later on. WDYT?

rottkit replied:
Since `options.html` is static, I would be ok assuming the element is always there.

bahadirk commented on popup.html:
```diff
@@ -0,0 +1,20 @@
+<div id="intro">
+    <img src="images/ados_test_listener96.png" style="float: right; height: 50px; width: 50px"/>
+<h1>Teamscale Listener for SAP Testing Events in the Azure DevOps</h1>
+    <p>Events are forwarded to Teamscale, to trigger the recording of coverage during test runs of manual SAP tests.
```
Small think but 
```suggestion
    <p>Events are forwarded to Teamscale to trigger the recording of coverage during test runs of manual SAP tests.
```

rottkit replied:
done

bahadirk commented on popup.js:
```diff
@@ -0,0 +1,137 @@
+		if (request.msg === tsSapCallFinishedMsg) {
+			const messagesElement = 'messages';
+			document.getElementById(messagesElement).innerHTML = formatServerResponse(request.data.serverResponse) +
+				document.getElementById(messagesElement).innerHTML;
```
I guess we can extract `document.getElementById(messagesElement)` as a local variable because we use it twice. Plus this would also allow us to check whether it returns an `undefined` or not and then to return early. WDYT?

rottkit replied:
done

bahadirk commented on popup.js:
```diff
@@ -0,0 +1,137 @@
+	let textParts = serverResponse.text.split('ManualSapTestInfo');
+	let text = textParts[0];
+	if (textParts.length > 1) {
+		text = textParts[0] +
```
```suggestion
		text += ...
```
?

bahadirk commented on popup.js:
```diff
@@ -0,0 +1,137 @@
+		labelsClasses += ' error';
+	}
+
+	return "<div class='" + labelsClasses + "'><div class='label'>" + serverResponse.action + "</div><div class='label'" +
```
I guess these divs and replace methods in the method make the code a bit difficult to read. How about defining a dictionary at the beginning of the method that consists the texts that are being defined and their values to be replaced? 
For instance 
```
const formatValues = {
"textKeyPrefix"  : {
    "value": "Test identifier: '", 
     "format": "'<span class="log-link" id="' + serverResponse.date + '-log" data- 
    test-key="'" 
},
"testKeyPostfix": {...}
...
}
```
I think this would help us to read the code easily, especially later on, and to see all the values in one place. WDYT?

rottkit replied:
Yeah... this one is definitely not nice. However I'm not quite sure about if improving is worth it (did not get your suggested change completely, I think). We could also have a quick chat about this one :-)

bahadirk replied:
What I was thinking in my mind is something like this starting from line 43:

```
...
const formatValues = {
	'lineBreak': {
		'old': /(?:\r\n|\r|\n)/g,
		'new': '<br>'
	},
	'resetUserStateUrl': {
		'old': 'api/projects/{project}/sap-test-event/reset/{sap-user}',
		'new': resetUserStateUrl + ' (<span class="reset-link" id="' + serverResponse.date + '-reset">click' +
			' here to send request now)</span>'
	},
	'testKeyPrefix': {
		'old': 'Test identifier: ',
		'new': '<span class="log-link" id="' + serverResponse.date + '-log" data-test-key="'
	},
	'testKeyPostfix': {
		'old': ' get log via api/projects/{project}/sap-test-event/log/{sap-test-key}.',
		'new': '">Click to retrieve/update log.</span>'
	},
	'link': {
		'old': ':/CQSE/MSG_TIA:000 ',
		'new': ''
	}
}

text = text.replace(formatValues['lineBreak']['old'], formatValues['lineBreak']['new']);
text = text.replace(formatValues['resetUserStateUrl']['old'], formatValues['resetUserStateUrl']['new']);
text = text.replace(formatValues['testKeyPrefix']['old'], formatValues['testKeyPrefix']['new']);
text = text.replace(formatValues['testKeyPostfix']['old'], formatValues['testKeyPostfix']['new']);
text = text.replace(formatValues['link']['old'], formatValues['link']['new']);

...
```
In the previous one, the HTML and JS codes seem a bit mingled with each other which might make the interpretation difficult. Here, it is a bit clear which values will be replaced by which value, I think. 
Again your call :) If you want to go with this approach, please double check the values one more because I quickly did the refactoring. And feel free to rename the variable names.

bahadirk commented on popup.js:
```diff
@@ -0,0 +1,137 @@
+
+function retainOnlyNewestOfSubsequentLogEntries(messagesElement) {
+	const messagesChildren = document.getElementById(messagesElement).children;
+	if (messagesChildren.length >= 4) {
```
Seems like it is a magic number, at least for me :) I think these magic numbers might be difficult to understand later on for those who take over the code later. I would probably suggest explaining it with an inline comment or directly in the function doc.

rottkit replied:
Added comment :-)

bahadirk commented on popup.js:
```diff
@@ -0,0 +1,137 @@
+chrome.runtime.onMessage.addListener(
+	function (request, sender, sendResponse) {
+		if (request.msg === tsSapCallFinishedMsg) {
+			const messagesElement = 'messages';
```
```suggestion
			const messagesElementId = 'messages';
```

rottkit replied:
deleted

bahadirk commented on popup.js:
```diff
@@ -0,0 +1,137 @@
+				document.getElementById(messagesElement).innerHTML;
+			const date = request.data.serverResponse.date;
+
+			retainOnlyNewestOfSubsequentLogEntries(messagesElement);
```
If we extract `document.getElementById(messagesElement)` as a local variable, then we can also directly pass it to this `retainOnlyNewestOfSubsequentLogEntries` method and don't need to call it `getElementById()` there

rottkit replied:
done

bahadirk commented on popup.js:
```diff
@@ -0,0 +1,137 @@
+function retainOnlyNewestOfSubsequentLogEntries(messagesElement) {
+	const messagesChildren = document.getElementById(messagesElement).children;
+	if (messagesChildren.length >= 4) {
+		if (messagesChildren[0].innerHTML.includes('Log</') && messagesChildren[2].innerHTML.includes('Log</')) {
```
maybe extracting `'Log</'` as a local variable?

rottkit replied:
extracted check to new function

bahadirk commented on popup.js:
```diff
@@ -0,0 +1,137 @@
+				}
+			});
+		}
+	};
```
This part and the one in the `initializeLogLink` method look very same to each other. Maybe extracting them as an additional method and passing them the necessary id and the data would be better. 

You can maybe even combine the entire`initializeResetLink` and `initializeLogLink` into one method. That would really increase reusability :)

rottkit replied:
done

bahadirk commented on popup.js:
```diff
@@ -0,0 +1,137 @@
+
+function makeObjectInspectionCollapsible(date) {
+	const infoElement = document.getElementById(date + '-info');
+	if (infoElement) {
+		listeners.push({elementId: infoElement.id, listener: () => triggerMoreInfo(date)});
+	}
+
+	const teaserElement = document.getElementById(date + '-teaser');
+	if (teaserElement) {
+		listeners.push({elementId: teaserElement.id, listener: () => triggerMoreInfo(date)});
+	}
```
AFAIS, In both these if blocks, the only thing changes is the `id` part. Maybe again extracting this listener part as a private method and pass the id and the date parameters to it?

rottkit replied:
Do you think this would improve readability?

bahadirk replied:
Would probably increase maintainability a bit because they seem to me like a potential code duplication candidate later on which we could maybe prevent in case this project gets bigger. 

But again just a small suggestion. You can also leave it as is :)

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+/* Map Chrome Tab Id to: Azure DevOps Session. This information currently unused. */
+let tabToSessionMap = {};
+/* Map Azure DevOps Session to: Azure DevOps User. This information is currently unused. */
+let sessionToUserMap = {};
```
For naming, in Dev team we generally tend to follow the following notation for map variables -> `...ById`. This would help us to understand what the key and the value is without looking the comment. 

For instance `usersBySessionId` or `testParametersById`. So you might also consider to follow this notation as well :)

rottkit replied:
done

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+const tsProjectOptionName = 'ts-project';
+const sapUserOptionName = 'sap-username';
+const extendedUriFilterOptionName = 'extended-uri-filter';
+const allOptions = [tsServerOptionName, tsProjectOptionName, sapUserOptionName, extendedUriFilterOptionName];
```
A similar suggestion like above -> `...Name` to `...Id`. 

Plus, since they seem like they are also defined in `options.js` maybe extracting them in a global place if I am not overseen something. Therefore whenever we want to change one of these ids we don't need to go through the files that use this id. WDYT?

rottkit replied:
done

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+const API_SERVICES_CALL_OPEN_TEST_RUNNER = '/_apis/Contribution/dataProviders/query';
+const API_CALL_UPDATE_TEST_RUN_SUFFIX = '/_api/_testresult/Update?teamId=&__v=5';
+
+const standardUriFilter = ["https://*.visualstudio.com/*", "https://dev.azure.com/*"];
```
```suggestion
const standardUriFilter = ['https://*.visualstudio.com/*', 'https://dev.azure.com/*'];
```
to adhere to our JS/TS coding guideline?

rottkit replied:
done

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+
+function cacheTeamscaleSessionCookie() {
+	const teamscaleServer = new URL(configOptions[tsServerOptionName]);
+	const setTeamscaleSessionCallback = function (cookieArray) {
+		for (const cookie of cookieArray) {
```
```suggestion
	const setTeamscaleSessionCallback = function (cookies) {
		for (const cookie of cookies) {
```

rottkit replied:
Since we have no type information at hand, I think in this case we can leave like that.

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+	const teamscaleServer = new URL(configOptions[tsServerOptionName]);
+	const setTeamscaleSessionCallback = function (cookieArray) {
+		for (const cookie of cookieArray) {
+			if (!(cookie.name.includes('teamscale-session') && cookie.name.includes('-' + teamscaleServer.port))) {
```
Probably extracting `cookie.name` as a local variable since we use it more than once?

rottkit replied:
Would only name it `cookieName`, so I do not see huge benefit from extracting it in this rather simple context.

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+		chrome.webRequest.onBeforeRequest.removeListener(listeners[tabId]);
+		delete listeners[tabId];
+	} else {
+		listeners[tabId] = (details) => testRunUpdateCallListener(details, tabId, tab);
```
```suggestion
		listeners[tabId] = details => testRunUpdateCallListener(details, tabId, tab);
```

rottkit replied:
done

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+);
+
+chrome.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
+	if (!changeInfo.status || changeInfo.status !== 'complete' ||
+		!(tab.title.startsWith('Runner') && tab.title.endsWith('Test Plans'))) {
```
Extracting `changeInfo.status` and `tab.title` as a local variable?

rottkit replied:
Since I would just name `tab.title` `tabTitle` I see no benefit of an extraction.

bahadirk replied:
Okay :)

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+	cacheTeamscaleSessionCookie();
+
+	initialListeners[tabId * 2] = details => resolveAzureDevOpsSessionIdListener(details, tabId);
+	initialListeners[tabId * 2 + 1] = details => onStartListener(details, tabId);
```
Some magic numbers again :) Since we use them also below, I would probably extract them as a local variable, give a meaningful name, and use them also below. Even an additional comment would be better, I guess

rottkit replied:
No magic number, just have two listeners per tab that are stored next to each other in the array. E.g. tabId=1 -> Listeners on position 2 and 4, tabId=5: listeners on 10 and 11.
Created two separate arrays to store the listeners :-)

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+		});
+
+		currentUriFilter = standardUriFilter;
+		if (configOptions[extendedUriFilterOptionName] && configOptions[extendedUriFilterOptionName].trim().length > 1) {
```
Extracting `configOptions[extendedUriFilterOptionName]` as a local variable?

rottkit replied:
done

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+
+function resolveUserNameOfTesterAndTriggerRecordingStart(apiUrl, tabId) {
+	const request = new XMLHttpRequest();
+	request.open("GET", apiUrl, true);
```
Since the HTTP methods like `GET` are common, I would probably extract them as a **global** enum, and replace all the usage of hardcoded `GET`s with it

rottkit replied:
I would say it's ok to have them inline. Or would you strictly prefer them to have in a variable?

bahadirk replied:
Again this is something we generally do in Dev Team. If something occurs more than once and can be also a global parameter, we try to extract it into a common place. 

So nothing critical here. Could be helpful to increase the maintainability a bit and that's all. If you say fine, then also fine to me :)

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+		if (request.readyState !== 4) {
+			return;
+		}
+		if (request.status === 200) {
```
Since the HTTP request codes like `200` are common, I would probably extract them as a **global** enum, and replace all the usage of hardcoded `200`s with it.

rottkit replied:
Will change

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+	const request = new XMLHttpRequest();
+	request.open("GET", apiUrl, true);
+	request.onreadystatechange = () => {
+		if (request.readyState !== 4) {
```
Again a magic number that might have an inline comment or explanation in the method doc :)

rottkit replied:
This is a request value of XMLHttpRequest, I added explanation link in the line above

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+		} else {
+			throw 'Could not obtain username.';
+		}
+	}
```
```suggestion
	};
```

rottkit replied:
fixed :-)

bahadirk commented on popup.js:
```diff
@@ -0,0 +1,137 @@
+});
+
+chrome.runtime.onMessage.addListener(
+	function (request, sender, sendResponse) {
```
Are `sender` and `sendResponse` unused parameters? Can we remove them?

rottkit replied:
I would prefer leaving it like that, to match the called method's interface :-)

bahadirk commented on popup.js:
```diff
@@ -0,0 +1,137 @@
+			const messagesElement = 'messages';
+			document.getElementById(messagesElement).innerHTML = formatServerResponse(request.data.serverResponse) +
+				document.getElementById(messagesElement).innerHTML;
+			const date = request.data.serverResponse.date;
```
Maybe extracting `request.data` as a local variable to increase reusability?

rottkit replied:
Can you elaborate on what you mean by "increase reusability"? Not sure whether the change would improve readability or fail-safety here.

bahadirk replied:
I generally tend to extract as many common parts as I can as a local variable. Because, if an option name (this is `data` here) changes, then we can only change it once where it defined, and not change all its occurrences by finding manually. Maybe in this context, it is not high likely possible, but I am speaking generally :)

Also, in case we need to call another option name on `request.data.serverResponse` later on, then we might not need to call the entire path once again, but only call the variable already defined for `request.data.serverResponse` or `request.data`. That was the reason behind `reusability`.

If I mixed the terms up with each other, sorry for that :) Like I said it is a very small suggestion, but again your call :)

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+		workItemId = JSON.parse(String.fromCharCode.apply(null, new Uint8Array(details.requestBody.raw[0].bytes))).context.properties.workItemIds;
+		apiCallUserInfoUri = details.url.substring(0, details.url.length - API_SERVICES_CALL_OPEN_TEST_RUNNER.length) +
+			'/_api/_common/GetUserProfile?__v=5';
+	}
```
Seems like the only part that this if-else block change is `API_ON_PREMISE_CALL_OPEN_TEST_RUNNER `, `API_SERVICES_CALL_OPEN_TEST_RUNNER `, and `jsonParse`. The rest is quite the same with each other. So maybe we can take the part that is not changed from the if-else block and only assign the values that are subject to change.

how about something like:

```
let calls;
let workItemId;
const jsonParse = JSON.parse(String.fromCharCode.apply(null, new Uint8Array(details.requestBody.raw[0].bytes)));
if (details.url.endsWith(API_ON_PREMISE_CALL_OPEN_TEST_RUNNER)) {
	workItemId = jsonParse.workItemIds;
	calls = API_ON_PREMISE_CALL_OPEN_TEST_RUNNER;
} else {
	workItemId = jsonParse.context.properties.workItemIds;
	calls = API_SERVICES_CALL_OPEN_TEST_RUNNER;
	
}
const apiCallUserInfoUri = details.url.substring(0, details.url.length - calls.length) +
	'/_api/_common/GetUserProfile?__v=5';
tabToTestCaseMap[tabId] = workItemId;
```

I just tried to refactor it quickly, I have also overseen something, but I guess you get the idea :)

rottkit replied:
refactored as suggested, thanks :-)

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+		return;
+	}
+
+	let updateRequest = JSON.parse(JSON.parse(String.fromCharCode.apply(
+		null, new Uint8Array(details.requestBody.raw[0].bytes))).updateRequests)[0];
```
What if we cannot parse it as JSON and then try to get the first element? Should we also check this case?

rottkit replied:
I would let the Extension fail which leads to an error in the Chrome extension overview and I think the only thing we could do when catching this problem would be to throw another exception.

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+
+	const testNameWithParameter = updateRequest.testCaseResult.testCaseTitle.trim();
+
+	if (updatedIterationActionResult.outcome === 12) { // test is paused, other call to Teamscale needed
+		queryTeamscale('pause', details.tabId, testNameWithParameter,
+			testOutcomeToTeamscaleTestExecutionResult(updatedIterationActionResult.outcome));
+	} else {
+		queryTeamscale('update', details.tabId, testNameWithParameter,
+			testOutcomeToTeamscaleTestExecutionResult(updatedIterationActionResult.outcome));
```
Here, maybe introducing a new local variable, setting it in if-else block and then calling the `queryTeamscale()` method?
```
let action;
if (updatedIterationActionResult.outcome === 12) {
 action = 'pause';
else {
 action = 'update';
}
queryTeamscale(action, details.tabId, testNameWithParameter,
			testOutcomeToTeamscaleTestExecutionResult(updatedIterationActionResult.outcome));
```

rottkit replied:
refactored as suggested, thanks :-)

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+			continue;
+		}
+
+		if (!parametersPerIteration[parameter.iterationId]) {
```
Maybe extracting `parametersPerIteration[parameter.iterationId]` as a local variable?

rottkit replied:
Since we also "write" to this variable, I think this is not possible (would require reassignment after write, right?)

bahadirk replied:
Fine by me, then :)

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+
+	const testId = tabToTestCaseMap[tabId];
+
+	if (action === 'update' && (!extendedName || !status)) {
```
Maybe extracting this `update` and `reset` values as a **global** enum? I don't know how we can do this in JS but I think it would be nice if we have one. Therefore we don't need to think too much about the places where these actions are being used and whether they are correctly defined

rottkit replied:
extracted all actions to a constant in `constants.js`

bahadirk commented on background.js:
```diff
@@ -0,0 +1,471 @@
+		}
+	}
+
+	text = text.replace(/ : .:\/CQSE\/MSG_TIA:000/gm, ':');
```
Maybe directly returning the `text.replace(/ : .:\/CQSE\/MSG_TIA:000/gm, ':')`?

rottkit replied:
done

bahadirk reviewed (commented):
I tried to review as much as I can since I am not an expert at writing chrome extensions :) That's why almost all of my comments are related to mostly formatting and a bit of reusability. I am also setting it to `Request Changes` but this is just for not leaving anything behind without a message, not because of that you have to apply all of them. 

So please feel free to say "we should keep it as it is because...." :)

rottkit committed:
CR#19156 Rework

rottkit committed:
CR#19156 Minor

rottkit committed:
CR#19156 Fix link initializer in popup

r2h2 committed:
FIx: Plugin breaks if no port is configured

The previous implementation assumed that the teamscaleServer configuration specifies a port.
In fact, the configured URL should not require a port. Instead, we guess the right cookie from the convention that TGA instance's ports begin with 8 and TQE instance's ports begin with 7.

r2h2 committed:
Check for cookie path

Simplify fix and check if cookie path is included in teamscaleServer.pathName

r2h2 committed:
Fix replacement of slashes

server pathname slashes at the end are stripped, as well as slashes at the beginning or end of cookies.

r2h2 committed:
use global search in JS regexp

Without the `g` option, only a leading OR trailing slash is replaced
